File: vehicle_gateway_python/examples/vtol_position_control.py
# Copyright 2023 Open Source Robotics Foundation, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import math
import sys
import time

import vehicle_gateway
from vehicle_gateway import ControllerType


def calc_distance_latlon(lat_1, lon_1, lat_2, lon_2):
    # This uses a planar approximation, not the real trigonometry. This
    # approximation is just fine for short distances, which is all we're
    # currently considering.
    EARTH_RADIUS = 6378100  # meters
    dx = (EARTH_RADIUS * math.pi / 180) * (lon_2 - lon_1) * math.cos(lat_1)
    dy = (EARTH_RADIUS * math.pi / 180) * (lat_2 - lat_1)
    return math.sqrt(dx * dx + dy * dy)


vg = vehicle_gateway.init(args=sys.argv, plugin_type='px4')
TARGET_ALTITUDE = 30  # meters above takeoff point

print('Arming...')
vg.arm_sync()
time.sleep(2)  # not sure why... perhaps some internal state setting?

print('Takeoff!')
home_lat, home_lon, home_alt = vg.get_latlon()
print(f'home_alt: {home_alt}')

vg.takeoff()
time.sleep(1)
vg.go_to_latlon(home_lat, home_lon, home_alt + TARGET_ALTITUDE)
while True:
    time.sleep(1)
    lat, lon, alt = vg.get_latlon()
    dalt = alt - home_alt
    print(f'takeoff climb, current altitude diff: {dalt}')
    if dalt > TARGET_ALTITUDE - 5:
        break  # close enough...

print('Transitioning to fixed-wing...')
vg.transition_to_fw_sync()
print(f'VTOL state: {vg.get_vtol_state().name}')

vg.go_to_latlon(home_lat, home_lon, home_alt + TARGET_ALTITUDE)
while True:
    time.sleep(1)
    lat, lon, alt = vg.get_latlon()
    dalt = alt - home_alt
    print(f'fw transition climbout, alt: {dalt}')
    # wait until we recover to close to our target altitude
    if dalt > TARGET_ALTITUDE - 2:
        break  # close enough...

print('begin transitioning to offboard control')
for t in range(0, 20):
    time.sleep(0.1)
    vg.set_local_position_setpoint(0, 0, -TARGET_ALTITUDE)
    vg.set_offboard_control_mode(ControllerType.POSITION)

vg.set_offboard_mode()

print('enabled position controller')
target_north = 200
target_east = 0
target_airspeed = 15

lap_count = 0
while True:
    current_ned = vg.get_local_position()
    dx = target_east - current_ned[1]
    dy = target_north - current_ned[0]

    vg.set_offboard_control_mode(ControllerType.POSITION)
    vg.set_local_position_setpoint(target_north, target_east, -TARGET_ALTITUDE)
    dist = math.sqrt(dx * dx + dy * dy)
    print(f'dist: {dist}')

    if dist < 10:
        print('changing target')
        target_north *= -1
        if target_north > 0:
            target_airspeed = 15  # fly slow towards the north
            lap_count += 1
            if lap_count >= 1:
                break
        else:
            target_airspeed = 20  # fly fast towards the south

    # I don't know why you have to repeatedly send it, but it seems necessary
    vg.set_airspeed(target_airspeed)
    time.sleep(0.1)

while True:
    vg.set_offboard_control_mode(ControllerType.POSITION)
    vg.set_local_position_setpoint(0, 0, -TARGET_ALTITUDE)
    time.sleep(0.1)
    current_ned = vg.get_local_position()
    dx, dy = -current_ned[1], -current_ned[0]
    dist = math.sqrt(dx * dx + dy * dy)
    print(f'dist: {dist}')
    if dist < 10:
        break

print('Switching back to hold mode...')
vg.set_onboard_mode()
vg.go_to_latlon(home_lat, home_lon, home_alt + TARGET_ALTITUDE)
time.sleep(5)

print('Transitioning to multicopter...')
vg.transition_to_mc_sync()
print(f'VTOL state: {vg.get_vtol_state().name}')

print('Hover above landing point...')
vg.set_onboard_mode()
vg.go_to_latlon(home_lat, home_lon, home_alt + TARGET_ALTITUDE)
for t in range(0, 60):
    time.sleep(1)
    vg.go_to_latlon(home_lat, home_lon, home_alt + TARGET_ALTITUDE)
    cur_lat, cur_lon, cur_alt = vg.get_latlon()
    distance = calc_distance_latlon(cur_lat, cur_lon, home_lat, home_lon)
    print(f't: {t} distance to home: {distance} alt: {cur_alt - home_alt}')
    if distance < 1:
        break

print('Descend to low altitude above landing point...')
vg.go_to_latlon(home_lat, home_lon, home_alt + 10)
for t in range(0, 30):
    time.sleep(1)
    cur_lat, cur_lon, cur_alt = vg.get_latlon()
    rel_alt = cur_alt - home_alt
    distance = calc_distance_latlon(cur_lat, cur_lon, home_lat, home_lon)
    print(f't: {t} distance to home: {distance} alt: {rel_alt}')
    if rel_alt < 12:
        break

print('Landing...')
vg.land()

# TODO(anyone): it would be nicer here to watch altitude or (even better)
# the descent rate, and break when it's ~zero
time.sleep(30)

print('Disarming...')
vg.disarm_sync()

vg.destroy()
print('Demo complete.')

File: vehicle_gateway_demo/src/vtol_position_control.cpp
// Copyright 2023 Open Source Robotics Foundation, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <iostream>

#include <pluginlib/class_loader.hpp>
#include <rclcpp/rclcpp.hpp>
#include <vehicle_gateway_px4/vehicle_gateway_px4.hpp>
#ifdef _WIN32
#include <Windows.h>
#else
#include <unistd.h>
#endif

class VehicleGatewayCpp
{
public:
  VehicleGatewayCpp()
  {
    this->loader_ = std::make_shared<pluginlib::ClassLoader<vehicle_gateway::VehicleGateway>>(
      "vehicle_gateway", "vehicle_gateway::VehicleGateway");
    this->gateway_ = this->loader_->createSharedInstance("vehicle_gateway_px4::VehicleGatewayPX4");
    this->gateway_->init(0, nullptr);
  }

  void destroy()
  {
    if (this->gateway_) {
      this->gateway_->destroy();
      this->gateway_ = nullptr;
    }
    if (this->loader_) {
      this->loader_->unloadLibraryForClass(
        "vehicle_gateway_px4::VehicleGatewayPX4");
      this->loader_ = nullptr;
    }
  }

  std::shared_ptr<vehicle_gateway::VehicleGateway> gateway_;

private:
  std::shared_ptr<pluginlib::ClassLoader<vehicle_gateway::VehicleGateway>> loader_;
};


int main(int argc, const char * argv[])
{
  rclcpp::init(argc, argv);

  const auto vg = std::make_shared<VehicleGatewayCpp>();
  const float TARGET_ATTITUDE = 30.0f;

  std::cout << "Arming..." << std::endl;
  vg->gateway_->arm_sync();
  sleep(2);

  const auto home_position = vg->gateway_->get_latlon();
  if (home_position.size() != 3) {
    throw std::runtime_error("home_position should have three elements: lat, lon, alt");
  }
  std::cout << "Home altitude: " << home_position[2] << std::endl;

  std::cout << "Takeoff!" << std::endl;
  vg->gateway_->takeoff();
  sleep(1);

  vg->gateway_->go_to_latlon_sync(
    home_position[0], home_position[1],
    home_position[2] + TARGET_ATTITUDE);

  vg->gateway_->transition_to_fw_sync();

  std::cout << "Begin transitioning to Offboard control..." << std::endl;
  vg->gateway_->transition_to_offboard_sync();

  std::cout << "Enabled position controller" << std::endl;
  auto target_north = 200.0,
    target_east = 0.0;

  std::cout << "Flying to first waypoint..." << std::endl;
  vg->gateway_->offboard_mode_go_to_local_setpoint_sync(
    target_north, target_east, -TARGET_ATTITUDE,
    std::numeric_limits<float>::quiet_NaN(), 15);
  std::cout << "Flying to second waypoint..." << std::endl;
  vg->gateway_->offboard_mode_go_to_local_setpoint_sync(
    -target_north, target_east, -TARGET_ATTITUDE,
    std::numeric_limits<float>::quiet_NaN(), 20);
  std::cout << "Flying home..." << std::endl;
  vg->gateway_->offboard_mode_go_to_local_setpoint_sync(
    0, 0, -TARGET_ATTITUDE,
    std::numeric_limits<float>::quiet_NaN(), 20);

  std::cout << "Switching back to hold mode..." << std::endl;
  vg->gateway_->set_onboard_mode();

  std::cout << "Transitioning to multicopter..." << std::endl;
  vg->gateway_->transition_to_mc_sync();
  std::cout << "VTOL state: " << vg->gateway_->get_vtol_state() << std::endl;
  sleep(1);

  vg->gateway_->go_to_latlon_sync(
    home_position[0], home_position[1],
    home_position[2] + TARGET_ATTITUDE);

  std::cout << "Landing..." << std::endl;
  vg->gateway_->land();

  std::cout << "Disarming..." << std::endl;
  vg->gateway_->disarm_sync();

  vg->destroy();
  std::cout << "Demo complete." << std::endl;

  rclcpp::shutdown();
  return 0;
}
